// eureca.io/examples/0-tests/client.js
var Eureca = require('../../');

var client = new Eureca.Client({ transport:'engine.io', uri: 'ws://localhost:8000/', /*prefix: 'eureca.io', retry: 3*/ });

client.ready(function (proxy) {

    proxy.hello();

    proxy.maths.add(10, 12.2).onReady(function (r) { console.log('10 + 12.2 = ', r) });

    
    proxy.maths.add(50, 100).onReady(function (r) { console.log('10+100 = ', r) });

    for (var i=1; i<3; i++)
    {
        for (var j=2; j<10; j+=2)
        {
            (function (i, j) {
                proxy.maths.add(i, j).onReady(function (r) { console.log(i+'+'+j+' = ', r) });
            })(i, j);
        }
    }
    
    //client.socket.send({__signal__:'signal data'});
});

client.exports.add = function (a, b) {
    return a + b;
}

client.exports.echo = function (msg) {
    console.log('ECHO = ', msg);
}

client.on('unhandledMessage', function (data) {
    console.log('unhandledMessage = ', data);
})

// Test zone =========================
/*
var http = require('http');
var querystring = require('querystring');


var post_data = querystring.stringify({
    __signal__: JSON.stringify({ data: 'foo' })
});

var post_options = {
    host: 'localhost',
    port: '8000',
    path: '/webrtc-eureca.io',
    method: 'POST',
    headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': post_data.length
    }
};
var post_req = http.request(post_options, function (res) {
    res.setEncoding('utf8');
    res.on('data', function (chunk) {
        console.log('Response: ' + chunk);
    });
});

post_req.write(post_data);
post_req.end();
console.log('sent post request', post_data);
*/
// Test zone =========================
